cache/client.go

```diff
@@ -0,0 +1,122 @@
+package cache
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/Sia/encoding"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/merkletree"
+)
+
+func verifyBlockHeader(header types.BlockHeader, minTimestamp types.Timestamp) error {
+	// Check that the timestamp is not too far in the past to be acceptable.
+	if minTimestamp > header.Timestamp {
+		return fmt.Errorf("Block header validation failed: EarlyTimestamp")
+	}
+
+	// Check if the block is in the extreme future. We make a distinction between
+	// future and extreme future because there is an assumption that by the time
+	// the extreme future arrives, this block will no longer be a part of the
+	// longest fork because it will have been ignored by all of the miners.
+	if header.Timestamp > types.CurrentTimestamp()+types.ExtremeFutureThreshold {
+		return fmt.Errorf("Block header validation failed: ExtremeFutureTimestamp")
+	}
+
+	// Check if the block is in the near future, but too far to be acceptable.
+	// This is the last check because it's an expensive check, and not worth
+	// performing if the earlier checks failed.
```

Comment:
I don't see how this check is more expensive than the previous one. Maybe we can remove the previous check if it always fails when this check fails?

Reply:
See the updates comment and my response above.